<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-0.9.449">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Data Infrastructure Framework Project - A framework for an open and scalable infrastructure for health data exemplified by the DD2 initiative</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link id="quarto-text-highlighting-styles" href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


</head>

<body class="floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Data Infrastructure Framework Project</span>
  </a>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://www.stenoaarhus.dk/research"><i class="bi bi-globe" role="img">
</i> 
 </a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/StenoAarhusRes"><i class="bi bi-twitter" role="img">
</i> 
 </a>
  </li>  
</ul>
              <div id="quarto-search" class="" title="Search"></div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto">
    <nav id="TOC" role="doc-toc">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#lay-description" id="toc-lay-description" class="nav-link active" data-scroll-target="#lay-description">Lay description</a></li>
  <li><a href="#summary" id="toc-summary" class="nav-link" data-scroll-target="#summary">Summary</a></li>
  <li><a href="#background" id="toc-background" class="nav-link" data-scroll-target="#background">Background</a></li>
  <li><a href="#methods" id="toc-methods" class="nav-link" data-scroll-target="#methods">Methods</a>
  <ul class="collapse">
  <li><a href="#overview-of-the-data-infrastructure-framework" id="toc-overview-of-the-data-infrastructure-framework" class="nav-link" data-scroll-target="#overview-of-the-data-infrastructure-framework">Overview of the data infrastructure framework</a>
  <ul class="collapse">
  <li><a href="#key-principles-underlying-the-framework" id="toc-key-principles-underlying-the-framework" class="nav-link" data-scroll-target="#key-principles-underlying-the-framework">Key principles underlying the framework</a></li>
  <li><a href="#selected-software-and-tools-to-use" id="toc-selected-software-and-tools-to-use" class="nav-link" data-scroll-target="#selected-software-and-tools-to-use">Selected software and tools to use</a></li>
  </ul></li>
  <li><a href="#the-frameworks-layers" id="toc-the-frameworks-layers" class="nav-link" data-scroll-target="#the-frameworks-layers">The framework’s layers</a>
  <ul class="collapse">
  <li><a href="#frontend-interface-the-web-portal" id="toc-frontend-interface-the-web-portal" class="nav-link" data-scroll-target="#frontend-interface-the-web-portal">Frontend interface: The web portal</a></li>
  <li><a href="#middle-layer-the-api" id="toc-middle-layer-the-api" class="nav-link" data-scroll-target="#middle-layer-the-api">Middle-layer: The API</a></li>
  <li><a href="#backend-the-database-and-documentation" id="toc-backend-the-database-and-documentation" class="nav-link" data-scroll-target="#backend-the-database-and-documentation">Backend: The database and documentation</a></li>
  </ul></li>
  <li><a href="#user-to-backend-pipeline" id="toc-user-to-backend-pipeline" class="nav-link" data-scroll-target="#user-to-backend-pipeline">User-to-backend pipeline</a>
  <ul class="collapse">
  <li><a href="#user-1-inputting-data" id="toc-user-1-inputting-data" class="nav-link" data-scroll-target="#user-1-inputting-data">User 1: Inputting data</a></li>
  <li><a href="#user-2-requesting-access" id="toc-user-2-requesting-access" class="nav-link" data-scroll-target="#user-2-requesting-access">User 2: Requesting access</a></li>
  <li><a href="#user-3-viewing-details" id="toc-user-3-viewing-details" class="nav-link" data-scroll-target="#user-3-viewing-details">User 3: Viewing details</a></li>
  <li><a href="#user-4-administrators-and-data-controllers" id="toc-user-4-administrators-and-data-controllers" class="nav-link" data-scroll-target="#user-4-administrators-and-data-controllers">User 4: Administrators and data controllers</a></li>
  </ul></li>
  <li><a href="#additional-key-components" id="toc-additional-key-components" class="nav-link" data-scroll-target="#additional-key-components">Additional key components</a>
  <ul class="collapse">
  <li><a href="#security-privacy-and-legality" id="toc-security-privacy-and-legality" class="nav-link" data-scroll-target="#security-privacy-and-legality">Security, privacy, and legality</a></li>
  <li><a href="#dissemination-of-framework" id="toc-dissemination-of-framework" class="nav-link" data-scroll-target="#dissemination-of-framework">Dissemination of framework</a></li>
  <li><a href="#sustainability" id="toc-sustainability" class="nav-link" data-scroll-target="#sustainability">Sustainability</a></li>
  <li><a href="#training-and-educational-development" id="toc-training-and-educational-development" class="nav-link" data-scroll-target="#training-and-educational-development">Training and educational development</a></li>
  </ul></li>
  <li><a href="#resource-requirements" id="toc-resource-requirements" class="nav-link" data-scroll-target="#resource-requirements">Resource requirements</a>
  <ul class="collapse">
  <li><a href="#computational-and-physical-infrastructure" id="toc-computational-and-physical-infrastructure" class="nav-link" data-scroll-target="#computational-and-physical-infrastructure">Computational and physical infrastructure</a></li>
  <li><a href="#technical-skills-and-personnel" id="toc-technical-skills-and-personnel" class="nav-link" data-scroll-target="#technical-skills-and-personnel">Technical skills and personnel</a></li>
  </ul></li>
  <li><a href="#applying-the-framework-to-dd2" id="toc-applying-the-framework-to-dd2" class="nav-link" data-scroll-target="#applying-the-framework-to-dd2">Applying the framework to DD2</a></li>
  <li><a href="#deliverables-and-milestones" id="toc-deliverables-and-milestones" class="nav-link" data-scroll-target="#deliverables-and-milestones">Deliverables and milestones</a></li>
  </ul></li>
  <li><a href="#collaborations" id="toc-collaborations" class="nav-link" data-scroll-target="#collaborations">Collaborations</a></li>
  <li><a href="#significance" id="toc-significance" class="nav-link" data-scroll-target="#significance">Significance</a></li>
  </ul>
<div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/steno-aarhus/dif-project/edit/main/index.Rmd" class="toc-action">Edit this page</a></p><p><a href="https://github.com/steno-aarhus/dif-project/issues/new" class="toc-action">Report an issue</a></p><p><a href="https://github.com/steno-aarhus/dif-project/blob/main/index.Rmd" class="toc-action">View source</a></p></div></div></nav>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">A framework for an open and scalable infrastructure for health data exemplified by the DD2 initiative</h1>
</div>



<div class="quarto-title-meta">

    
    
  </div>
  

</header>

<div id="fig-anonymous-1" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><a href="https://doi.org/10.5281/zenodo.6511112"><img src="https://zenodo.org/badge/DOI/10.5281/zenodo.6511112.svg" class="img-fluid figure-img" data-align="left"></a></p>
</figure>
</div>
<section id="lay-description" class="level1">
<h1>Lay description</h1>
<p>Many initiatives in clinical research, including the Danish Centre for Strategic Research in Type 2 Diabetes (DD2) initiative, have a hard time getting funds for building and maintaining needed software that make sure health science is the highest quality it can be. For this project, we aim to build software tools that make it easier to do better research, especially for managing and working with data. We’ll first be building these tools to help support the DD2 initiative.</p>
<p>We will be sharing these tools widely and freely, so that as many research groups as possible can use it for their own projects. Not only will we build these tools to help researchers with managing and sharing their data, we also will create them with beginner-friendly documentation and training material to make sure more researchers use our tools, no matter their skill level. We believe that with these tools, science on health and well-being can become better, ultimately helping people with diabetes and society in general.</p>
</section>
<section id="summary" class="level1">
<h1>Summary</h1>
<p>In clinical and health research, especially for small- to mid-sized research groups, funding for building modern, open source software infrastructures for managing and using data has been limited. This gap has naturally led to organizational challenges for managing existing and incoming data for many research initiatives, including the Danish Centre for Strategic Research in Type 2 Diabetes (DD2) initiative, a national research collaboration and database initiative established in 2010 with continual enrollment of persons with type 2 diabetes. The aim of our project is to close this gap by creating and implementing an efficient, scalable, and open source data infrastructure framework that connects data collectors, researchers, clinicians, and other stakeholders, with the data, documentation, and findings within the DD2 study. This will improve and extend the existing DD2 research infrastructure into an open national state-of-the-art research infrastructure that will provide easy and transparent access to this resource for researchers, clinicians and stakeholders, thus enabling excellent data science driven research. Furthermore, we will create this framework in such a way that other research groups and companies, who are unable to adequately invest in building infrastructures of this type, can relatively easily implement it, and modify as needed, for their own purposes. By building this framework, we have the potential to help propel research groups and companies across Denmark (and globally) to quickly getting updated on modern, scalable, and efficient approaches to working with data. Within the DD2 setting, an open, transparent, and easy access to this constantly growing resource has the potential of greatly improving the interest in, use of, and scientific impact of this resource, thus leading to substantial scientific and medical advancements, individualised treatment and improved human health in not only persons with type 2 diabetes, but population overall.</p>
</section>
<section id="background" class="level1">
<h1>Background</h1>
<p>In clinical research, software and data infrastructure development is undervalued and, aside from this funding call, underfunded, particularly for small- to mid-sized research organizations. Clinical and health researchers largely lack formal training, support, and awareness in research software engineering (RSE) and in building and managing data infrastructures. The result is that the overall software and computational ecosystem, as well as the technical capacities to maintain them, lags behind multiple other scientific domains (e.g., bioinformatics). Particularly with the recent rise of data science and the greater focus on analytical reproducibility, this issue has become increasingly apparent as data, and the skills required to work with it, become ever larger, more technical, and complex. Indeed, investing in and implementing scalable and modern data infrastructures and RSE processes, built on open source software, have the potential to greatly improve the quality of science, to produce more transparent and streamlined workflows, to lead to reproducible research, and generally better science in less time <span class="citation" data-cites="Lowndes2017">(<a href="#ref-Lowndes2017" role="doc-biblioref">1</a>)</span>.</p>
<p>Funding for participant recruitment and data acquisition has historically been (and still is) easier to obtain than for building open source software and infrastructures that support and enhance science, particularly for managing and using data. This imbalance has naturally led to organizational challenges for managing existing and incoming data for many research initiatives within the field of clinical research, including for the Danish Centre for Strategic Research in Type 2 Diabetes (DD2) initiative <span class="citation" data-cites="Nielsen2012 DD2">(<a href="#ref-Nielsen2012" role="doc-biblioref">2</a>,<a href="#ref-DD2" role="doc-biblioref">3</a>)</span>.</p>
<p>DD2 is a national type 2 diabetes (T2D) research collaboration and database initiative that was established in 2010, with on-going enrollment by hospital physicians and general practitioners (GPs). Although T2D is a single diagnosis, it comprises several phenotypes with diverse prognoses and risks for complications, which can lead to treatments tailored to each phenotype. The overarching aim of DD2 is to improve and individualise the treatment of persons with T2D. Figure 1 shows the several datasets within DD2 <span class="citation" data-cites="Jakobsen2021 Gylfadottir2020 Stidsen2017 Valentiner2019">(<a href="#ref-Jakobsen2021" role="doc-biblioref">4</a>–<a href="#ref-Valentiner2019" role="doc-biblioref">7</a>)</span>. DD2 has received extensive funding from the Danish Council for Strategic Research and the Novo Nordisk Foundation as well as a Steno National Collaborative Grant for deep phenotyping. Continuously recruiting more participants, adding new data, and expanding the data access to researchers throughout Denmark and abroad has the potential to further increase the value of DD2. However, this comes with higher costs and resources for maintaining, extending, and improving on the existing DD2 research infrastructure.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/dd2-description.svg" class="img-fluid figure-img"></p>
<p></p><figcaption aria-hidden="true" class="figure-caption">Figure 1: Summary of the different DD2 databases.</figcaption><p></p>
</figure>
</div>
<p>Building modern data infrastructures has slowly been taking greater priority by funding and research agencies globally. For instance, the UK Biobank <span class="citation" data-cites="Sudlow2015 UKBiobank">(<a href="#ref-Sudlow2015" role="doc-biblioref">8</a>,<a href="#ref-UKBiobank" role="doc-biblioref">9</a>)</span> is a large-scale biomedical database with highly detailed data on ~500,000 participants. It is regularly expanded with additional data and is globally accessible to approved researchers and is a role model to building a modern research infrastructure.</p>
<p>While the UK Biobank is a source of inspiration on the state-of-the-art, the underlying infrastructure itself is not openly accessible and reusable. The same applies to a similar Danish initiative, the “<em>Single path to access Danish health data</em>” project <span class="citation" data-cites="sundhedsdata">(<a href="#ref-sundhedsdata" role="doc-biblioref">10</a>)</span>, where the Danish government and individual regions are collaborating to map out all Danish health data. Another state-of-the-art initiative led by the University of Chicago, USA is Gen3 <span class="citation" data-cites="gen3">(<a href="#ref-gen3" role="doc-biblioref">11</a>)</span>, which contains modular open source services that can form the basis for a data infrastructure <span class="citation" data-cites="ctds gen3software">(<a href="#ref-ctds" role="doc-biblioref">12</a>,<a href="#ref-gen3software" role="doc-biblioref">13</a>)</span> and powers several research platforms, including the National Institutes of Health <span class="citation" data-cites="gen3platform">(<a href="#ref-gen3platform" role="doc-biblioref">14</a>)</span>. However, we are unaware of any similar current national efforts that are open source, re-usable, and suitable for the Danish and EU legal context.</p>
<p>Our primary aim is to create and implement an efficient, scalable, and open source data infrastructure <em>framework</em> that connects data collectors, researchers, clinicians, and other stakeholders, with the data, documentation, and findings within the DD2 study. This will improve and extend the existing DD2 into an open national state-of-the-art research infrastructure that will provide easy and transparent access to this resource for researchers, thus enabling excellent data science driven research. Our secondary aim is to create this framework in such a way so that other research groups and companies, who are unable to adequately invest in building similar infrastructures, can relatively easily implement it and modify as needed for their own purposes.</p>
</section>
<section id="methods" class="level1">
<h1>Methods</h1>
<p>Our first steps are to build a data infrastructure framework and the second step is to implement it in DD2. The details of the framework itself are described first and later we describe how we will apply it to DD2.</p>
<section id="overview-of-the-data-infrastructure-framework" class="level2">
<h2 class="anchored" data-anchor-id="overview-of-the-data-infrastructure-framework">Overview of the data infrastructure framework</h2>
<p>For this project, the data infrastructure <em>framework</em> is defined as 1) a set of software programs, 2) a defined and fixed set of conventions on the structure and format of the filesystem and URL paths, and 3) a defined structure to the data and associated documentation, all of which are linked together as modular components. The framework will serve as an open source starting template for setting up data infrastructures that make use of modern tools and processes.</p>
<p>This framework encompasses four target users and three layers, with a complete schematic shown in Figure 2. The three layers are the web portal frontend, the database and documentation backend, and the API (Application Programming Interface) that interacts with both. The four users and their associated use cases are:</p>
<ol type="1">
<li><strong>User 1</strong>: Those <em>inputting data</em>, e.g., authorized centers and researchers. The use cases are:
<ol type="1">
<li>Authorized centers (GPs and hospitals) upload standardized and routine data through the data entry web portal.</li>
<li>Authorized researchers upload generated data from completed projects. Non-standardized data is manually processed and cleaned before entering into the database.</li>
</ol></li>
<li><strong>User 2</strong>: Those <em>requesting access</em>, e.g., researchers and clinicians. Use cases are:
<ol type="1">
<li>Interested researchers browse the catalogue of available data and the data dictionary.</li>
<li>Researchers request access to data by submitting a description of their proposed project and selecting the relevant data from the catalogue. This request is sent to a list of projects to await approval from the data controllers (User 4).</li>
</ol></li>
<li><strong>User 3</strong>: Those <em>viewing updates on findings and results</em> such as aggregate statistics, e.g., policymakers, healthcare workers, journalists, researchers, and the general public. Use cases are:
<ol type="1">
<li>Users view and read through the list of completed, ongoing, and proposed projects that use the database.</li>
<li>Users access and view aggregate statistics and the latest published findings that are relevant to them/their practice.</li>
</ol></li>
<li><strong>User 4</strong>: Administrators and data controllers. Use cases are:
<ol type="1">
<li>Approve newly submitted projects that request data and manage existing projects.</li>
<li>Approve and authorize researchers to access the web portal for data requests and to manage current authorized researchers.</li>
<li>Manage collection centers’ access and authorization.</li>
</ol></li>
</ol>
<p>Throughout this application, we’ll refer to these four users and three layers as we expand on and describe the framework.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<figure class="figure">
<img src="images/detailed-schematic.svg" title="Schematic of DIF." class="img-fluid figure-img">
</figure>
<p></p><figcaption aria-hidden="true" class="figure-caption">Figure 2: Detailed schematic of the data infrastructure framework.</figcaption><p></p>
</figure>
</div>
<section id="key-principles-underlying-the-framework" class="level3">
<h3 class="anchored" data-anchor-id="key-principles-underlying-the-framework">Key principles underlying the framework</h3>
<p>To ensure the development of this framework is efficient and focused, it will adhere to key principles that are supported by strong philosophical and scientific rationale:</p>
<ol type="1">
<li>Follow FAIR (Findable, Accessible, Interoperable, and Reusable), open, and transparent principles for the framework itself and to enable FAIR principles in the database.</li>
<li>Be openly licensed and re-usable to facilitate uptake in other national groups (or companies) who are unable to invest in and build it.</li>
<li>Uses state-of-the-art principles and tools in software design and development.</li>
<li>Be beginner-friendly and targeted to (potential) non-computationally technical users of the framework.</li>
</ol>
<p>In order to maximise the potential for re-use and to minimise the technical debt and expertise needed to use, maintain, and modify the framework, we will use software and tools underlying the framework that fit these principles:</p>
<ol type="1">
<li>Wherever possible, re-use existing material: There already exists many great software tools, infrastructures, and resources that haven’t been incorporated into common health research practice. We will make use of and/or modify these materials where we can.</li>
<li>Be familiar to or used by researchers currently or within the near future: To ensure the greatest potential for continued maintenance, development, and use, the framework should use or be built with tools and skills that are at least familiar or soon to be familiar to researchers.</li>
<li>Be familiar to skilled personnel (e.g.&nbsp;research software engineers, data engineers, data scientists): Skilled personnel will build this framework and need to be familiar with them.</li>
<li>Be open source: Software that isn’t open source is by definition not transparent, FAIR, or open. This is a requirement as it will encourage wide and easy re-use.</li>
<li>Integrates easily with other software: Modular software that follows common input/output conventions and has well-designed and documented APIs are easier to build with and maintain.</li>
<li>Historically stable and reliable: While there are always new software being built, maintenance and development is easier when using those that are established.</li>
<li>Likely to be used in the future or is easily interchangeable with potential future tools: Technology progresses quite quickly, so we will rely on software that is likely to still be used or can be switched to other tools.</li>
</ol>
</section>
<section id="selected-software-and-tools-to-use" class="level3">
<h3 class="anchored" data-anchor-id="selected-software-and-tools-to-use">Selected software and tools to use</h3>
<p>Based on the above principles, we have chosen the following software and conventions to form the framework’s foundation:</p>
<p>&lt;&lt;&lt;&lt;&lt;&lt;&lt; Updated upstream - <strong>R and/or Python:</strong> Are open source and established. - <strong>RESTful API:</strong> REST (representational state transfer) is the de facto standard for building APIs. - <strong>CSV, SQL, and JSON format for data storage:</strong> Are open source and established formats for data and other web information. - JSON (JavaScript Object Notation) is the Internet’s de facto standard format for information transfer. - CSV is well known to researchers. - SQL (a querying language and data storage format) is a widely used format for storing data. - <strong>Git, GitLab, and DVC for version control: </strong>Formal version control (VC) systems are software that track changes to files and are standard practice in software engineering. After Git was developed, it has increasingly been used by researchers across scientific fields to improve research transparency and openness. GitLab is an open source product to host Git projects and handles many modern software development tasks. DVC, or Data Version Control, is a machine learning VC software for data and models that integrates with Git. - <strong>Markdown for documentation:</strong> Markdown is a common web format and is quickly becoming a standard in data science and even within many scientific fields. It is a simple syntax for writing documents that enables easy conversion to file formats like HTML or Word. It is simple and portable, so will be used for writing the data dictionary, results, documentation, and training material. - <strong>Tools for modern web-based interfaces:</strong> These include HTML, CSS, JavaScript, and other web technologies, as well as User Interface (UI)/User Experience (UX) design principles. - <strong>Software development best practices: </strong>This includes unit testing, continuous integration, and document-driven development (to emphasize usability). We’ll use the concept of Minimum Viable Product (MVP) as a means of quickly building something that is minimally workable. - <strong>Re-use of similar infrastructures:</strong> Modules from initiatives like Gen3 will either be re-used directly or modified to fit our requirements. ======= - <strong>R and/or Python:</strong> Are open source and established. - <strong>RESTful API:</strong> REST (representational state transfer) is the de facto standard for building APIs. - <strong>CSV, SQL, and JSON format for data storage:</strong> Are open source and established formats for data and other web information. - JSON (JavaScript Object Notation) is the Internet’s de facto standard format for information transfer. - CSV is well known to researchers. - SQL (a querying language and data storage format) is a widely used format for storing data. - <strong>Git, GitLab, and DVC for version control:</strong> Formal version control (VC) systems are software that track changes to files and are standard practice software engineering. After Git was developed, it has increasingly been used by researchers across scientific fields to improve research transparency and openness. GitLab is an open source product to host Git projects and handles many modern software development tasks. DVC, or Data Version Control, is a machine learning VC software for data and models that integrates with Git. - <strong>Markdown for documentation:</strong> Markdown is a common web format and is quickly becoming a standard in data science and even within many scientific fields. It is a simple syntax for writing documents that enables easy conversion to file formats like HTML or Word. It is simple and portable, so will be used for writing the data dictionary, results, documentation, and training material. - <strong>Tools for modern web-based interfaces:</strong> These include HTML, CSS, JavaScript, and other web technologies, as well as User Interface (UI)/User Experience (UX) design principles. - <strong>Software development best practices:</strong> This includes unit testing, continuous integration, and document-driven development (to emphasize usability). We’ll use the concept of Minimum Viable Product (MVP) as a means of quickly building something that is minimally workable. - <strong>Re-use of similar infrastructures:</strong> Modules from initiatives like Gen3 will either be re-used directly or modified to fit our requirements. &gt;&gt;&gt;&gt;&gt;&gt;&gt; Stashed changes</p>
</section>
</section>
<section id="the-frameworks-layers" class="level2">
<h2 class="anchored" data-anchor-id="the-frameworks-layers">The framework’s layers</h2>
<section id="frontend-interface-the-web-portal" class="level3">
<h3 class="anchored" data-anchor-id="frontend-interface-the-web-portal">Frontend interface: The web portal</h3>
<p>This interface is what all users interact with and use, with essentially three “permission” levels available:</p>
<ol type="1">
<li>Full access: User 4.</li>
<li>Authorized access:
<ol type="1">
<li>User 1: A data upload portal that accepts either the routine, known data or new, unknown data from User 2.</li>
<li>User 2: A data request portal that allows the user to select variables in the catalogue of data as well as write and submit project proposals for access.</li>
</ol></li>
<li>Public access: User 3 would have access to all public pages, which includes the data dictionary, updates on findings, list of current and past projects, and a log of any changes or additions to the data.</li>
</ol>
<p>All content would be rendered directly as plain HTML text to ease use of existing webpage translation services (e.g.&nbsp;Google Translate), so that content written in another language, i.e., Danish, would still be readable to non-native speakers. This would also lower the amount of maintenance necessary for documentation.</p>
</section>
<section id="middle-layer-the-api" class="level3">
<h3 class="anchored" data-anchor-id="middle-layer-the-api">Middle-layer: The API</h3>
<p>Modern web and computational infrastructures are built on web APIs. Any modern online resource or interface makes use of an API, such as from Google, Gen3, or the UK Biobank. An API is a mechanism by which different programs can communicate with one another. They form a set of instructions or conventions that allow easy communication between a user and the computer. APIs by their nature are transparent and if well-documented would ensure the linked data would be FAIR, safely and securely.</p>
<p>In this case, the API would be between the user and the web server that stores the underlying database and documentation. The API would be a combination of a predefined set of instructions that are sent to the web server to run certain commands as well as a set of explicit conventions and rules on how files and folders are structured and named. Taken together, this API would allow other software like R packages to be built to interact with the backend to automate tasks done by the users.</p>
</section>
<section id="backend-the-database-and-documentation" class="level3">
<h3 class="anchored" data-anchor-id="backend-the-database-and-documentation">Backend: The database and documentation</h3>
<p>Given the heterogeneity in the sources of data input, the backend will need to be composed of multiple components: raw data files as plain text, cleaning and processing programming scripts, a formal database structure (e.g.&nbsp;SQL), a VC system to track changes to the raw data and processing scripts, a data version numbering system, a changelog describing the changes, and a data dictionary linked to the variables contained in the database. Versioning of the raw data and scripts is done for recordkeeping, auditing, and transparency, in addition to allowing comparison of data used between past and current projects that use the data.</p>
<p>A major challenge to building the backend is in the heterogeneity of the data input. The key is to establish and enforce a standardized Common Data Model (CDM) for all incoming data at the point of entry. For the framework, the exact contents of the database aren’t important, since as long as the contents follow the CDM it will be programmatically merged into the final formal database. This is necessary as the database contents depend heavily on the research topic and aims of the study that will use this framework.</p>
<p>The backend documentation is either largely generated automatically or manually written. For instance, the list of projects and findings would be generated by the submitted projects and input from User 2 (researchers) while the changelog would be updated either by automated additions or, optionally, manually from User 4. The data dictionary would be stored as a JSON file with the documentation text itself as Markdown text. This data dictionary would be publicly accessible and could be updated by anyone (with approval from User 4), potentially through “Merge Request” mechanisms. This mechanism involves automatically linking any addition or correction back to the main documentation and requesting it be merged into it.</p>
</section>
</section>
<section id="user-to-backend-pipeline" class="level2">
<h2 class="anchored" data-anchor-id="user-to-backend-pipeline">User-to-backend pipeline</h2>
<section id="user-1-inputting-data" class="level3">
<h3 class="anchored" data-anchor-id="user-1-inputting-data">User 1: Inputting data</h3>
<p>Depending on the source of data, there may already be established data input processes. Substantial amounts of biomedical data, especially in Denmark, come from already established, routinely collected clinical data such as from outpatient clinics. For these sources of data, the data input pipeline would involve redirecting these data sources through the API and storage format so the data continues on to the backend.</p>
<p>Sources of data that don’t have well-established data input processes, such as from hospitals, medical laboratories, and so on, would use the data input portal. This portal would only accept data that is in a pre-defined format and would include documentation, and potentially automation scripts, on how to pre-process the data prior to uploading it.</p>
<p>Once the data is submitted through the portal, it would get sent in an encrypted, legally-compliant format to the server and stored in the way defined by the API and CDM. Any new or updated data that is uploaded would trigger generic automated data cleaning, processing, quality control checks of this new data. Any automated processing that is developed specific to a project would need to adhere to the API’s conventions. If any issues are found or if the data is entirely new to the database, they get sent to a log and User 4 would receive a notification to deal with the issue. If there are no issues or the issues have been dealt with, an automated script would take a snapshot of the data with the VC system, the version number (based on <a href="https://semver.org/">Semantic Versioning</a>) of the data would be updated, an entry would get added to the changelog, and the formal database would get updated.</p>
</section>
<section id="user-2-requesting-access" class="level3">
<h3 class="anchored" data-anchor-id="user-2-requesting-access">User 2: Requesting access</h3>
<p>Researchers and other users who want to request access to the data would first need their identity verified and then be approved for authorized access. After approval, they would interact with the frontend by two routes:</p>
<p>&lt;&lt;&lt;&lt;&lt;&lt;&lt; Updated upstream 1. Browsing from the catalogue of existing variables in the database, along with the data dictionary and the data changelog, and then selecting those variables that are relevant to their project. The catalogue of variables would be a table that is built from the underlying set of files that contain the documentation and that includes a button to select and save the variables of interest. If they desired, they could also update any errors or typos in the data dictionary through the “Merge Request” mechanism described above. 2. Writing a description of the scientific rationale of the project and adding any relevant information into a web form that User 4 would use to decide on approval and that would be displayed on the website for User 3. ======= 1. Browsing from the catalogue of existing variables in the database, along with the data dictionary and the data changelog, and then selecting those variables that are relevant to their project. The catalogue of variables would be a table that is built from the underlying set of files that contain the documentation and that includes a button to select and save the variables of interest. If they desired, they could also update any errors or types in the data dictionary through the “Merge Request” mechanism described above. 2. Writing a description of the scientific rationale of the project and adding any relevant information into a web form that User 4 would use to decide on approval and that would be displayed on the website for User 3. &gt;&gt;&gt;&gt;&gt;&gt;&gt; Stashed changes</p>
<p>When User 4 approves a data request project, it will trigger an API request that would automatically extract the requested subset of data, bundle and encrypt it, and send it to the researcher’s secure server. The framework will contain sufficiently generic methods for automating the data transfer process.</p>
</section>
<section id="user-3-viewing-details" class="level3">
<h3 class="anchored" data-anchor-id="user-3-viewing-details">User 3: Viewing details</h3>
<p>The framework assumes that this user would interact with the portal through at least three routes:</p>
<ol type="1">
<li>Reading about the database, its history, organizational structure and ownership, and any other details in a typical “About” section. Since this is a part of a basic website layout, it is not fully part of the framework.</li>
<li>Browsing completed and ongoing projects making use of the database, along with the description of the project and variables used. This would be based on a “database” of projects that have requested access to the data, would be stored in JSON format, and through the API specification would be generated into a table on the portal.</li>
<li>Getting updates on the latest results from the projects and browsing some basic aggregate statistics of some key variables. This would involve programmatically generating standard aggregate statistics of the underlying database, formatting results according to the API, transferring them to the web server, and have it generated into a webpage format. Overlapping with the second route, key findings from completed projects can be stored in the projects database and be generated into a listing on the website.</li>
</ol>
</section>
<section id="user-4-administrators-and-data-controllers" class="level3">
<h3 class="anchored" data-anchor-id="user-4-administrators-and-data-controllers">User 4: Administrators and data controllers</h3>
<p>These users would largely interact with the web portal for managing and overseeing ongoing projects, approve access for new researchers, and approve projects requesting access to data. Approving new researchers would grant the researchers access permissions to enter the User 2 portal.</p>
</section>
</section>
<section id="additional-key-components" class="level2">
<h2 class="anchored" data-anchor-id="additional-key-components">Additional key components</h2>
<section id="security-privacy-and-legality" class="level3">
<h3 class="anchored" data-anchor-id="security-privacy-and-legality">Security, privacy, and legality</h3>
<p>The framework itself does not contain any personal data. When the framework is deployed as an infrastructure for a database, only aggregate statistics, and not individual-level personal data, would be publicly accessible. Any personal data would be stored on a secure server that is decided and controlled by User 4, who would be responsible for complying with relevant legal requirements.</p>
<p>For data transfers of personal data, either from data collection centers, data generated from researchers, or when transferring data for approved projects, we would use well-established and compliant encrypted data transfer processes. Key authentication principles such as two-factor authentication and OAuth (open standard for access authentication) will be central to the framework to control who can update or transfer the data. The endpoint of the data transfer is dealt with by the legal teams of the relevant institutions.</p>
</section>
<section id="dissemination-of-framework" class="level3">
<h3 class="anchored" data-anchor-id="dissemination-of-framework">Dissemination of framework</h3>
<p>To be aligned to the goals of openness, transparency, and FAIR principles, the complete development of the framework will take place openly on GitHub. From there we will link to and promote it through various outlets, including publications, conferences, and social media. The framework and all its components will be licensed under permissive copyright licenses like the MIT License for the software and the Creative Commons Attribution License for non-software content.</p>
</section>
<section id="sustainability" class="level3">
<h3 class="anchored" data-anchor-id="sustainability">Sustainability</h3>
<p>Integral to this framework is ensuring it is sustainable over the long-term by:</p>
<ol type="1">
<li>Integrating the development of the framework and using it in multiple projects at the host institution Steno Diabetes Center Aarhus (SDCA).</li>
<li>Prototyping MVP to build and document maintenance procedures.</li>
<li>Incorporating fundamental open source sustainability concepts and documentation such as writing a Code of Conduct and Contributing Guidelines and encouraging external users to contribute.</li>
<li>Including duties for the hired personnel to contribute back to open source software that the framework depends on to help their, and indirectly the framework’s, sustainability.</li>
</ol>
</section>
<section id="training-and-educational-development" class="level3">
<h3 class="anchored" data-anchor-id="training-and-educational-development">Training and educational development</h3>
<p>Usage of this framework depends on the quality of its documentation and training material. A key concept we will use heavily is Documentation Driven Development, where the framework’s development is guided and informed from the development of the documentation, which places documentation as a high priority. We will also be creating and running short workshops and tutorials that teach researchers how they can use this framework.</p>
</section>
</section>
<section id="resource-requirements" class="level2">
<h2 class="anchored" data-anchor-id="resource-requirements">Resource requirements</h2>
<section id="computational-and-physical-infrastructure" class="level3">
<h3 class="anchored" data-anchor-id="computational-and-physical-infrastructure">Computational and physical infrastructure</h3>
<p>While the proposed framework is software-based, storing and deploying the infrastructure requires server space and IT support. The framework itself takes up little space and can optimise computational resources, but the underlying DD2 data requires considerable server space.</p>
</section>
<section id="technical-skills-and-personnel" class="level3">
<h3 class="anchored" data-anchor-id="technical-skills-and-personnel">Technical skills and personnel</h3>
<p>To have a meaningful impact on improving research infrastructure, the minimum skills and knowledge necessary are:</p>
<ul>
<li>Building and interacting with RESTful APIs (or APIs in general)</li>
<li>Creating websites and knowledge on UI/UX designs</li>
<li>Developing software/general programming in R/Python</li>
<li>Using Git and GitHub/GitLab</li>
<li>Database management/administration</li>
<li>Familiar with creating training material and teaching</li>
<li>Experience in working on open source projects and with FAIR principles</li>
<li>Experience in the Danish research computing environments, e.g., “forskermaskine” (research engine on Denmark Statistics (DST))</li>
</ul>
</section>
</section>
<section id="applying-the-framework-to-dd2" class="level2">
<h2 class="anchored" data-anchor-id="applying-the-framework-to-dd2">Applying the framework to DD2</h2>
<p>The biggest potential challenge to applying the framework to DD2 is getting the database backend into the appropriate structure to fit within the framework. With the current state of the DD2 data, considerable time and effort is needed to organize it. Our initial steps will be to:</p>
<ol type="1">
<li><em>Survey and map out all data, documentation, and processing steps</em>. Currently, the original enrollment data are stored in a secure server at the DD2 headquarters at Odense, while subsets of data collected for specific research projects are spread across several research institutions.</li>
<li><em>Map the data input sources and formats from the various collection centers</em>. GP clinics have an existing pipeline for sending data to DD2 through their system, “sundhedsdatanettet” (national healthcare portal), following a patient visit. Other collection centers like hospitals have custom, but not standardized, approaches to sending data to DD2. No formal approach is available for returning data from completed projects. The core DD2 data is sent to “forskermaskine” and gets merged there with the registry data, which is in a standard DST format.<br>
</li>
<li><em>Move as much data as possible to a central location</em>. All data will be stored at the DD2 headquarters in Odense, except large-scale data that will be either stored or transferred as needed to a high-performance computing (HPC) platform. For linkage with registry data, we will acquire a dedicated DD2 “forskermaskine” server.</li>
<li><em>Re-structure current data into the framework’s CDM</em>.</li>
<li><em>Build software to automate the cleaning, processing, and merging of the existing and established data input pipelines into the framework’s required backend format</em>.</li>
<li><em>Establish automated processes for linkages between the data storage servers</em>.</li>
<li><em>Implement framework’s remaining modules, starting with User 4 and then moving from User 1 to 3, in that order</em>.</li>
</ol>
<p>Currently, User 3 can request data by filling out a Word <a href="https://dd2.dk/media/1410/standard-dd2-protocol_final.doc">application form</a> and emailing it to the chair of the advisory board Kurt Højlund and programme leader Jens Steen Nielsen. Applications are reviewed by the steering committee and, once approved, the data manager at the Department of Clinical Epidemiology (KEA) in Aarhus University Hospital then manually extracts the requested data and transfers the data subset to the applicant’s secure server and does this for each individual research project. If requested, KEA may also perform analyses on the data. Researchers must already have valid authorized access to the secure servers on an existing “forskermaskine” or an <a href="https://www.deic.dk/en/supercomputing/national-hpc-facilities">HPC facility</a> for the large-scale data, such as <a href="https://www.computerome.dk/">Computerome 2</a> and <a href="https://genome.au.dk/">GenomeDK</a>.</p>
<p>The costs of storing the original data are covered by DD2, while applicants cover the costs related to storing the transferred data. We will not charge for data access. As per legal requirements, researchers can only use the data for the intended purposes listed in the application. After project completion, the researchers must delete or close access to the data and inform DD2 as legally required. Any newly generated data must be returned to DD2 by uploading via the User 1 portal.</p>
</section>
<section id="deliverables-and-milestones" class="level2">
<h2 class="anchored" data-anchor-id="deliverables-and-milestones">Deliverables and milestones</h2>
<p>Because the framework will be built with modularity in mind, where each component could be used alone or together, nearly all the components could be deliverables (each User by each layer). Each deliverable would be to prototype a MVP to begin testing it, identifying bugs, getting feedback, and establishing maintenance procedures. See Figure 3 for the Gantt chart.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<figure class="figure">
<img src="images/gantt-chart.svg" title="Gantt Chart" class="img-fluid figure-img">
</figure>
<p></p><figcaption aria-hidden="true" class="figure-caption">Figure 3: Gantt chart of deliverables and timelines.</figcaption><p></p>
</figure>
</div>
</section>
</section>
<section id="collaborations" class="level1">
<h1>Collaborations</h1>
<p>The framework will be developed at SDCA with Professor Annelli Sandbæk (applicant) as the lead PI responsible for reaching the overall goals of the project and defined milestones, as well as two postdocs, Luke Johnston, MSc, PhD and Alisa Kjærgaard, MD, PhD. A project group headed by the PI will be established that includes central persons from SDCA and DD2. In close collaboration with the project manager of DD2 and the current data manager, the deliverables will be planned and carried through. Completing the proposed project requires hiring data and research software engineer personnel into the project, which will be the first step in the project process. The DD2 advisory group will also act as the advisory group for this project. This group is chaired by Kurt Højlund, MD, PhD, head of research at Steno Diabetes Center Odense and contains representatives from affiliated research projects and other DD2 stakeholders.</p>
</section>
<section id="significance" class="level1">
<h1>Significance</h1>
<p>We are at a key point in time within clinical research where it is increasingly being recognized that open source software and computational infrastructure are critical and necessary components to ensuring science is high-quality, reproducible, rigorous, and transparent. Funding agencies and research institutions globally are putting greater efforts into modernizing many of their infrastructures using the many software technologies that have risen in the last decade. By building this framework, we have the potential to help propel research groups and companies across Denmark (and globally) to quickly getting updated on modern, scalable, and efficient approaches to working with data.</p>
<p>Within the DD2 setting, an open, transparent, and easy access to this constantly growing resource has the potential of greatly improving the interest in, use of, and scientific impact of this resource. Incorporating new data generated from the DD2 resource back into DD2 will enable other researchers to test or use this data to advance their own work. This would lead to substantial scientific and medical advancements, which will ultimately lead to individualised treatment and improved human health in individuals with type 2 diabetes, and very likely the population overall.</p>



</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body" role="doc-bibliography">
<div id="ref-Lowndes2017" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">1. </div><div class="csl-right-inline">Lowndes JSS, Best BD, Scarborough C, Afflerbach JC, Frazier MR, O’Hara CC, et al. Our path to better science in less time using open data science tools. Nature Ecology &amp; Evolution [Internet]. 2017 May [cited 2021 May 2];1(6):1–7. Available from: <a href="https://www.nature.com/articles/s41559-017-0160">https://www.nature.com/articles/s41559-017-0160</a></div>
</div>
<div id="ref-Nielsen2012" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">2. </div><div class="csl-right-inline">Nielsen JS, Thomsen RW, Steffensen C, Christiansen JS. The danish centre for strategic research in type 2 diabetes (DD2) study: Implementation of a nationwide patient enrollment system. Clinical Epidemiology [Internet]. 2012 Sep [cited 2021 May 2];4(Supplement 1 Diabetes):27–36. Available from: <a href="https://www.dovepress.com/the-danish-centre-for-strategic-research-in-type-2-diabetes-dd2-study–peer-reviewed-fulltext-article-CLEP">https://www.dovepress.com/the-danish-centre-for-strategic-research-in-type-2-diabetes-dd2-study–peer-reviewed-fulltext-article-CLEP</a></div>
</div>
<div id="ref-DD2" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">3. </div><div class="csl-right-inline">DD2 [Internet]. DD2. [cited 2021 May 2]. Available from: <a href="https://dd2.dk/">https://dd2.dk/</a></div>
</div>
<div id="ref-Jakobsen2021" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">4. </div><div class="csl-right-inline">Jakobsen PR, Christensen JR, Nielsen JB, Søndergaard J, Ejg Jarbøl D, Olsen MH, et al. Identification of important factors affecting use of digital individualised coaching and treatment of type 2 diabetes in general practice: A qualitative feasibility study. International Journal of Environmental Research and Public Health [Internet]. 2021 Jan [cited 2021 May 2];18(8):3924. Available from: <a href="https://www.mdpi.com/1660-4601/18/8/3924">https://www.mdpi.com/1660-4601/18/8/3924</a></div>
</div>
<div id="ref-Gylfadottir2020" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">5. </div><div class="csl-right-inline">Gylfadottir SS, Christensen DH, Nicolaisen SK, Andersen H, Callaghan BC, Itani M, et al. <a href="https://doi.org/10.1097/j.pain.0000000000001744">Diabetic polyneuropathy and pain, prevalence, and patient characteristics: A cross-sectional questionnaire study of 5,514 patients with recently diagnosed type 2 diabetes</a>. Pain. 2020 Mar;161(3):574–83. </div>
</div>
<div id="ref-Stidsen2017" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">6. </div><div class="csl-right-inline">Stidsen JV, Nielsen JS, Henriksen JE, Friborg SG, Thomsen RW, Olesen TB, et al. Protocol for the specialist supervised individualised multifactorial treatment of new clinically diagnosed type 2 diabetes in general practice (IDA): A prospective controlled multicentre open-label intervention study. BMJ Open [Internet]. 2017 Dec [cited 2021 May 2];7(12):e017493. Available from: <a href="https://bmjopen.bmj.com/content/7/12/e017493">https://bmjopen.bmj.com/content/7/12/e017493</a></div>
</div>
<div id="ref-Valentiner2019" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">7. </div><div class="csl-right-inline">Valentiner LS, Thorsen IK, Kongstad MB, Brinkløv CF, Larsen RT, Karstoft K, et al. Effect of ecological momentary assessment, goal-setting and personalized phone-calls on adherence to interval walking training using the InterWalk application among patients with type 2 diabetes—a pilot randomized controlled trial. PLOS ONE [Internet]. 2019 Jan [cited 2021 May 2];14(1):e0208181. Available from: <a href="https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0208181">https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0208181</a></div>
</div>
<div id="ref-Sudlow2015" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">8. </div><div class="csl-right-inline">Sudlow C, Gallacher J, Allen N, Beral V, Burton P, Danesh J, et al. UK biobank: An open access resource for identifying the causes of a wide range of complex diseases of middle and old age. PLOS Medicine [Internet]. 2015 Mar [cited 2021 May 2];12(3):e1001779. Available from: <a href="https://journals.plos.org/plosmedicine/article?id=10.1371/journal.pmed.1001779">https://journals.plos.org/plosmedicine/article?id=10.1371/journal.pmed.1001779</a></div>
</div>
<div id="ref-UKBiobank" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">9. </div><div class="csl-right-inline">UK biobank [Internet]. [cited 2021 May 2]. Available from: <a href="https://www.ukbiobank.ac.uk/">https://www.ukbiobank.ac.uk/</a></div>
</div>
<div id="ref-sundhedsdata" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">10. </div><div class="csl-right-inline">En indgang til sundhedsdata [Internet]. [cited 2021 May 2]. Available from: <a href="https://www.enindgangtilsundhedsdata.dk/">https://www.enindgangtilsundhedsdata.dk/</a></div>
</div>
<div id="ref-gen3" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">11. </div><div class="csl-right-inline">Gen3 data commons [Internet]. [cited 2021 May 2]. Available from: <a href="http://gen3.org/">http://gen3.org/</a></div>
</div>
<div id="ref-ctds" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">12. </div><div class="csl-right-inline">Center for translational data science GitHub repositories [Internet]. GitHub. [cited 2021 May 2]. Available from: <a href="https://github.com/uc-cdis">https://github.com/uc-cdis</a></div>
</div>
<div id="ref-gen3software" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">13. </div><div class="csl-right-inline">Gen3 software [Internet]. Center for Translational Data Science. [cited 2021 May 2]. Available from: <a href="https://ctds.uchicago.edu/gen3">https://ctds.uchicago.edu/gen3</a></div>
</div>
<div id="ref-gen3platform" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">14. </div><div class="csl-right-inline">Gen3 - powered by Gen3 [Internet]. [cited 2021 May 2]. Available from: <a href="http://gen3.org/powered-by-gen3/">http://gen3.org/powered-by-gen3/</a></div>
</div>
</div></section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      let href = ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
      <div class="nav-footer-center">
        <ul class="footer-items list-unstyled">
    <li class="nav-item">
    <a class="nav-link" href="./LICENSE.html">License under CC-BY 4.0</a>
  </li>  
</ul>
      </div>
  </div>
</footer>



</body></html>